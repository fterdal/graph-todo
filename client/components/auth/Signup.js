import React, { Component } from 'react';
import { graphql, compose } from 'react-apollo';
import gql from 'graphql-tag';
import { meQuery } from '../SideBar';
import { Redirect } from 'react-router-dom';

/***** GraphQL *****/
const signupMutation = gql`
  mutation SignupMutation($email: String!, $password: String!){
    signup(
      input: {
        email: $email,
        password: $password,
      }
    ) {
      id
      email
    }
  }
`;

/***** React Component *****/
export class Signup extends Component {
  state = {
    email: '',
    password: '',
    confirmPassword: '',
    signedUp: false,
    errorMessage: '',
  }

  _signup = async () => {
    const { email, password, confirmPassword } = this.state
    if (password !== confirmPassword) {
      this.setState({
        password: '',
        confirmPassword: '',
        errorMessage: 'Password and Confirm Password must match',
      })
      return;
    }
    await this.props.signupMutation({
      variables: {
        email,
        password
      },
      refetchQueries: [
        { query: meQuery },
      ],
    })
    this.setState({ signedUp: true });
  }

  render() {
    const { _signup, state: {
      email,
      password,
      confirmPassword,
      signedUp,
      errorMessage,
    } } = this;
    if (signedUp) return (<Redirect to="/" />)
    const errorMsg = errorMessage ? <div className="container errorMsg">{errorMessage}</div>
                            : null
    return (
      <div className="container">
        <h1>Signup</h1>
        {errorMsg}
        <form>
          <fieldset>
            <label>Email</label>
            <input
              id="emailField"
              type="text"
              value={email}
              onChange={(evt) => this.setState({ email: evt.target.value })}
            />
            <label>Password</label>
            <input
              id="passwordField"
              type="password"
              value={password}
              onChange={(evt) => this.setState({ password: evt.target.value })}
            />
            <label>Confirm Password</label>
            <input
              id="confirmPasswordField"
              type="password"
              value={confirmPassword}
              onChange={(evt) => this.setState({ confirmPassword: evt.target.value })}
            />
            <a
              className="button auth-button"
              onClick={() => _signup()}>
              Signup
            </a>
          </fieldset>
        </form>
      </div>
    )
  }
}

/***** Apollo Wrapper *****/
export default compose(
  graphql(signupMutation, {
    props: ({ data, mutate }) => ({ data, signupMutation: mutate }),
  }),
)(Signup)
