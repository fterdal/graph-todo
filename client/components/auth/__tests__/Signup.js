/* eslint-env jest */
import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { Signup } from '../Signup';

test('Signup renders a form to sign up', () => {
  const signup = shallow(
    <Signup />
  );
  expect(signup).toMatchSnapshot();
});

test('User can enter in the signup fields', () => {
  const signupMutationMock = jest.fn();
  const signup = shallow(
    <Signup signupMutation={signupMutationMock} />
  );
  signup.find('#emailField').simulate('change', {
    target: { value: 'hello@world.com' }
  });
  signup.find('#passwordField').simulate('change', {
    target: { value: 'secretPassword' }
  });
  signup.find('#confirmPasswordField').simulate('change', {
    target: { value: 'secretPassword' }
  });
  expect(signup.state().email).toBe('hello@world.com');
  expect(signup.state().password).toBe('secretPassword');
  expect(signup.state().confirmPassword).toBe('secretPassword');
  expect(signup).toMatchSnapshot();
  signup.find('.auth-button').simulate('click');
  expect(signupMutationMock).toHaveBeenCalled();
});
