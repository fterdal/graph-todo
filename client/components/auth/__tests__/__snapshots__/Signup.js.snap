// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup renders a form to sign up 1`] = `
<div
  className="container"
>
  <h1>
    Signup
  </h1>
  <form>
    <fieldset>
      <label>
        Email
      </label>
      <input
        id="emailField"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label>
        Password
      </label>
      <input
        id="passwordField"
        onChange={[Function]}
        type="password"
        value=""
      />
      <label>
        Confirm Password
      </label>
      <input
        id="confirmPasswordField"
        onChange={[Function]}
        type="password"
        value=""
      />
      <a
        className="button auth-button"
        onClick={[Function]}
      >
        Signup
      </a>
    </fieldset>
  </form>
</div>
`;

exports[`User can enter in the signup fields 1`] = `
<div
  className="container"
>
  <h1>
    Signup
  </h1>
  <form>
    <fieldset>
      <label>
        Email
      </label>
      <input
        id="emailField"
        onChange={[Function]}
        type="text"
        value="hello@world.com"
      />
      <label>
        Password
      </label>
      <input
        id="passwordField"
        onChange={[Function]}
        type="password"
        value="secretPassword"
      />
      <label>
        Confirm Password
      </label>
      <input
        id="confirmPasswordField"
        onChange={[Function]}
        type="password"
        value="secretPassword"
      />
      <a
        className="button auth-button"
        onClick={[Function]}
      >
        Signup
      </a>
    </fieldset>
  </form>
</div>
`;
